File: c/statistics.c
/*
 * statistics.c - mean, median and mode of a list of integers.
 *
 * Procedural implementation for MSCS-632. The program is one translation unit
 * built from free functions that transform data passed to them explicitly;
 * there are no objects and no closures. Storage is a heap array that the
 * caller owns and must release, since C has no garbage collector.
 *
 * Build: gcc -Wall -Wextra -std=c11 -o statistics statistics.c
 * Usage: ./statistics [int ...]      (with no arguments a sample set is used)
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <limits.h>

/* Sample data used when the caller supplies no arguments. Deliberately
   bimodal so the mode routine has to return more than one value. */
static const int SAMPLE_DATA[] = {4, 1, 2, 2, 3, 5, 4, 2, 7, 4};
static const size_t SAMPLE_COUNT = sizeof(SAMPLE_DATA) / sizeof(SAMPLE_DATA[0]);

/*
 * Insertion sort, written out rather than delegated to qsort so that the
 * ordering logic is visible. O(n^2), which is irrelevant at these sizes and
 * keeps the control flow explicit.
 */
static void sort_ints(int *values, size_t n)
{
	for (size_t i = 1; i < n; i++) {
		int key = values[i];
		size_t j = i;
		while (j > 0 && values[j - 1] > key) {
			values[j] = values[j - 1];
			j--;
		}
		values[j] = key;
	}
}

/* Mean as a double so the fractional part is not silently truncated. */
static double compute_mean(const int *values, size_t n)
{
	long long sum = 0;   /* wider than int: 10^5 ints near INT_MAX would overflow */
	for (size_t i = 0; i < n; i++)
		sum += values[i];
	return (double)sum / (double)n;
}

/*
 * Median of an array that is already sorted. With an even count there is no
 * single middle element, so the two straddling values are averaged.
 */
static double compute_median(const int *sorted, size_t n)
{
	if (n % 2 == 1)
		return (double)sorted[n / 2];
	return ((double)sorted[n / 2 - 1] + (double)sorted[n / 2]) / 2.0;
}

/*
 * Mode of a sorted array. Because equal values are adjacent after sorting,
 * one pass counts each run and records the longest.
 *
 * A set may be multimodal, so the result is an array. It is allocated here and
 * becomes the caller's responsibility to free. *mode_count receives its length
 * and *frequency the winning run length.
 */
static int *compute_modes(const int *sorted, size_t n, size_t *mode_count, int *frequency)
{
	/* At most n distinct values, so this is always large enough. */
	int *modes = malloc(n * sizeof(int));
	if (modes == NULL) {
		perror("malloc");
		exit(EXIT_FAILURE);
	}

	size_t found = 0;
	int best = 0;

	size_t i = 0;
	while (i < n) {
		/* Walk to the end of the run of values equal to sorted[i]. */
		size_t run_start = i;
		while (i < n && sorted[i] == sorted[run_start])
			i++;
		int run_length = (int)(i - run_start);

		if (run_length > best) {
			/* Strictly better: discard everything collected so far. */
			best = run_length;
			found = 0;
			modes[found++] = sorted[run_start];
		} else if (run_length == best) {
			/* Ties are also modes. */
			modes[found++] = sorted[run_start];
		}
	}

	*mode_count = found;
	*frequency = best;
	return modes;
}

/*
 * Parse one argument into an int, rejecting junk and out-of-range values.
 * strtol is used rather than atoi because atoi cannot report failure.
 */
static int parse_int(const char *text, int *out)
{
	char *end = NULL;
	errno = 0;
	long value = strtol(text, &end, 10);

	if (end == text || *end != '\0')
		return 0;                       /* not a number, or trailing junk */
	if (errno == ERANGE || value > INT_MAX || value < INT_MIN)
		return 0;                       /* outside the range of int */

	*out = (int)value;
	return 1;
}

static void print_list(const char *label, const int *values, size_t n)
{
	printf("%-16s[", label);
	for (size_t i = 0; i < n; i++)
		printf("%d%s", values[i], (i + 1 < n) ? ", " : "");
	printf("]\n");
}

int main(int argc, char **argv)
{
	size_t n;
	int *values = NULL;

	if (argc > 1) {
		n = (size_t)(argc - 1);
		values = malloc(n * sizeof(int));
		if (values == NULL) {
			perror("malloc");
			return EXIT_FAILURE;
		}
		for (size_t i = 0; i < n; i++) {
			if (!parse_int(argv[i + 1], &values[i])) {
				fprintf(stderr, "error: '%s' is not a valid integer\n", argv[i + 1]);
				free(values);       /* release before the early exit */
				return EXIT_FAILURE;
			}
		}
	} else {
		n = SAMPLE_COUNT;
		values = malloc(n * sizeof(int));
		if (values == NULL) {
			perror("malloc");
			return EXIT_FAILURE;
		}
		memcpy(values, SAMPLE_DATA, n * sizeof(int));
		printf("No input given; using the built-in sample set.\n\n");
	}

	/* An empty data set has no mean, median or mode; say so rather than
	   dividing by zero. */
	if (n == 0) {
		fprintf(stderr, "error: no values to summarise\n");
		free(values);
		return EXIT_FAILURE;
	}

	print_list("Input:", values, n);

	/* Sorting in place is fine because the input order is not needed again.
	   Median and mode both require sorted data. */
	sort_ints(values, n);
	print_list("Sorted:", values, n);

	printf("\nCount:          %zu\n", n);
	printf("Mean:           %.4f\n", compute_mean(values, n));
	printf("Median:         %.4f\n", compute_median(values, n));

	size_t mode_count = 0;
	int frequency = 0;
	int *modes = compute_modes(values, n, &mode_count, &frequency);

	printf("Mode:           ");
	for (size_t i = 0; i < mode_count; i++)
		printf("%d%s", modes[i], (i + 1 < mode_count) ? ", " : "");
	printf("  (each occurring %d time%s", frequency, frequency == 1 ? "" : "s");
	if (mode_count > 1)
		printf("; the set is multimodal");
	printf(")\n");

	/* Every allocation above is released here. */
	free(modes);
	free(values);
	return EXIT_SUCCESS;
}

File: ocaml/statistics.ml
(* statistics.ml - mean, median and mode of a list of integers.

   Functional implementation for MSCS-632. Every value here is immutable:
   there are no references, no mutable record fields, and no loops. The whole
   program is expressions composed out of higher-order list functions
   (fold_left, map, filter, sort), and each statistic is a pure function of its
   argument.

   Build: ocamlfind ocamlopt -package str statistics.ml -o statistics
     or simply: ocaml statistics.ml
   Usage: ./statistics [int ...]      (with no arguments a sample set is used) *)

(* Deliberately bimodal, so [modes] has to return more than one value. *)
let sample_data = [ 4; 1; 2; 2; 3; 5; 4; 2; 7; 4 ]

(* Mean. Returns an option because the empty list has no mean; making that
   explicit in the type stops a caller from ignoring the case. *)
let mean (values : int list) : float option =
  match values with
  | [] -> None
  | _ ->
      let total = List.fold_left ( + ) 0 values in
      let count = List.length values in
      Some (float_of_int total /. float_of_int count)

(* Median. Sorting returns a new list rather than reordering the argument, so
   the caller's list is untouched. With an even count the two middle values are
   averaged. *)
let median (values : int list) : float option =
  match values with
  | [] -> None
  | _ ->
      let sorted = List.sort compare values in
      let n = List.length sorted in
      let mid = n / 2 in
      if n mod 2 = 1 then Some (float_of_int (List.nth sorted mid))
      else
        let lower = float_of_int (List.nth sorted (mid - 1)) in
        let upper = float_of_int (List.nth sorted mid) in
        Some ((lower +. upper) /. 2.0)

(* Frequency table as an association list, accumulated with a fold. The
   accumulator is rebuilt on each step rather than updated in place, which is
   what makes this usable without any mutable state. *)
let frequencies (values : int list) : (int * int) list =
  List.fold_left
    (fun table x ->
      match List.assoc_opt x table with
      | Some count -> (x, count + 1) :: List.remove_assoc x table
      | None -> (x, 1) :: table)
    [] values

(* Mode. A set may be multimodal, so this returns every value tied for the
   highest frequency, together with that frequency.

   The pipeline reads directly as the definition: build the table, find the
   largest count, keep the entries matching it, discard the counts, sort. *)
let modes (values : int list) : int list * int =
  match frequencies values with
  | [] -> ([], 0)
  | table ->
      let best = List.fold_left (fun acc (_, count) -> max acc count) 0 table in
      let winners =
        table
        |> List.filter (fun (_, count) -> count = best)
        |> List.map fst
        |> List.sort compare
      in
      (winners, best)

(* ---- presentation ---- *)

let string_of_int_list (values : int list) : string =
  "[" ^ String.concat ", " (List.map string_of_int values) ^ "]"

let describe_float (label : string) (value : float option) : string =
  match value with
  | None -> Printf.sprintf "%-15s undefined (empty list)" label
  | Some v -> Printf.sprintf "%-15s %.4f" label v

let describe_modes (values : int list) : string =
  match modes values with
  | [], _ -> Printf.sprintf "%-15s undefined (empty list)" "Mode:"
  | winners, frequency ->
      Printf.sprintf "%-15s %s  (each occurring %d time%s%s)" "Mode:"
        (String.concat ", " (List.map string_of_int winners))
        frequency
        (if frequency = 1 then "" else "s")
        (if List.length winners > 1 then "; the set is multimodal" else "")

(* Parse the command line into a list of ints. The fold carries a Result, so
   the first bad argument short-circuits the rest without an exception. *)
let parse_args (args : string list) : (int list, string) result =
  List.fold_left
    (fun acc text ->
      match (acc, int_of_string_opt text) with
      | Error e, _ -> Error e
      | Ok xs, Some v -> Ok (v :: xs)
      | Ok _, None -> Error text)
    (Ok []) args
  |> Result.map List.rev

let run (values : int list) : unit =
  print_endline (Printf.sprintf "%-15s %s" "Input:" (string_of_int_list values));
  print_endline
    (Printf.sprintf "%-15s %s" "Sorted:"
       (string_of_int_list (List.sort compare values)));
  print_newline ();
  print_endline (Printf.sprintf "%-15s %d" "Count:" (List.length values));
  print_endline (describe_float "Mean:" (mean values));
  print_endline (describe_float "Median:" (median values));
  print_endline (describe_modes values)

let () =
  let args = List.tl (Array.to_list Sys.argv) in
  if args = [] then (
    print_endline "No input given; using the built-in sample set.\n";
    run sample_data)
  else
    match parse_args args with
    | Ok values -> run values
    | Error bad ->
        prerr_endline (Printf.sprintf "error: '%s' is not a valid integer" bad);
        exit 1
